### PR TITLE
[Config] Handle nullable node name + fix inheritdocs

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no-ish
| Deprecations? | no
| Tests pass?   | yes    
| Fixed tickets | #...   
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

Small split from #26297 that can be merged until master/4.1. Whereas the doc fixes only apply until 3.4, hence the split.

Small change regarding `getName/Path()` for not returning a `null` value anymore which violates `NodeInterface::getName/Path()`

Remainng issue left at 

https://github.com/symfony/symfony/blob/cd5f4105a43208632c8b62fd75b4dad53dcd8a41/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L381-L383

which i tend to leave untouched across all branches for now.

### CODE REVIEW

**Comment by ro0NL** [comment]
Added all fixes that can be merged until master. Ill change #26297 to backport `string|null` fixes soley (which are type declarations as of 4.0>

File: src/Symfony/Component/Config/Definition/BaseNode.php [hunk 1]
```diff
@@ -40,7 +40,7 @@
      */
     public function __construct($name, NodeInterface $parent = null)
     {
-        if (false !== strpos($name, '.')) {
+        if (null !== $name && false !== strpos($name, '.')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Definition/BaseNode.php]: why not cast to string instead?
`if (false !== strpos($name = (string) $name, '.')) {`

Comment (ro0NL) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: yes, better. Now updated.

Comment (ro0NL) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: Didnt reassign `$name` btw... we need to preserve string|null on the property side.

Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: > we need to preserve string|null on the property side

why? would need a side effect, which one is it? casting all the time is hurting perf, that's not the side effect I'd consider good ;)

Comment (ro0NL) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: > which one is it?

BC... (`protected $name;`). Also the constructor sig kinda implies a nullable value.

Then again.. if i got this right, it shouldnt hurt indeed. AFAIK prototyped nodes will have a string name (`setName()`) once being dealt with. And the calling side needs to expect string anyway due `getName()`.

Given `$arr[null] = .. ` equals `$arr[''] = ...` (see remaining issue) i think it's safe to make the change :+1:

Comment (nicolas-grekas) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: > the constructor sig kinda implies a nullable value

not the docblock, isn't it? I'd support doing the change :)

Comment (ro0NL) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: touche. It does in 4.0 though. Now updated.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by ro0NL** [comment]
found a few more, derived from NodeInterface/PrototypeNodeInterface :sweat_smile:

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/Config/Definition/NodeInterface.php [hunk 2]
```diff
@@ -83,6 +89,9 @@ public function merge($leftSide, $rightSide);
      * @param mixed $value The value to finalize
      *
      * @return mixed The finalized value
+     *
+     * @throws Exception
```
Comment (fabpot) [inline src/Symfony/Component/Config/Definition/NodeInterface.php]: All `@throws` should have a description explaining when the exception can be triggered... or be removed :)

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Config/Definition/NodeInterface.php [hunk 3]
```diff
@@ -59,11 +63,13 @@ public function hasDefaultValue();
     public function getDefaultValue();
 
     /**
-     * Normalizes the supplied value.
+     * Normalizes a value.
      *
-     * @param mixed $value The value to normalize
+     * @param mixed $value Value to normalize
```
Comment (Simperfit) [inline src/Symfony/Component/Config/Definition/NodeInterface.php]: maybe `A value to normalize` ?

**Review by Simperfit — APPROVED** [review]

File: src/Symfony/Component/Config/Definition/NodeInterface.php (hunk 3 above)
Comment (ro0NL) [reply src/Symfony/Component/Config/Definition/NodeInterface.php]: in fact.. i think i should revert this one as the interface was already using different syntax:

https://github.com/symfony/symfony/pull/26335/files#diff-41913ce7a681852774440044e2c091bdR92

File: src/Symfony/Component/Config/Definition/BaseNode.php [hunk 4]
```diff
@@ -40,7 +40,7 @@
      */
     public function __construct($name, NodeInterface $parent = null)
     {
-        if (false !== strpos($name, '.')) {
+        if (false !== strpos($name = (string) $name, '.')) {
```
Comment (fabpot) [inline src/Symfony/Component/Config/Definition/BaseNode.php]: `$name` must be a string, so casting here should be removed. Adding this is not a bug fix.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Config/Definition/BaseNode.php (hunk 4 above)
Comment (ro0NL) [reply src/Symfony/Component/Config/Definition/BaseNode.php]: in 4.0 it's `string|null` already

https://github.com/symfony/symfony/blob/d65c43ba843c84105c194d9de03945b9cf5cee79/src/Symfony/Component/Config/Definition/BaseNode.php#L42

hence #26297

practically using `null` is just an _alias_ for empty string, but we use null by design: 

https://github.com/symfony/symfony/blob/d65c43ba843c84105c194d9de03945b9cf5cee79/src/Symfony/Component/Config/Definition/Builder/ArrayNodeDefinition.php#L75-L78

if we preserve `null` on the property side (thus no cast) we keep violating `getName`:

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Config/Definition/BaseNode.php#L223-L225

hence the cast

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @ro0NL.

**Pull request closed without merging**